---
title: |-
  BIOL 812 \
    Intro to Computation \
    and Big Data \
    in Biology
author: "Rob Colautti"
output: html_document
---

NOTE: Tutorials run Wednesdays 12:30-2:30 in BioSci 1120

# Readings & Assignments {.tabset .tabset-fade .tabset-pills} 

<span style="color:red"> Weekly assignments will appear here </span>

Last Updated:
`r format(Sys.time(), '%d %B, %Y')` 

## Week 1

Review [Fundamental R](https://colauttilab.github.io/RCrashCourse/1_fundamentals.html) and make sure you are comfortable with the content.

If you think you are comfortable with the content, then take the [self-assessment](https://colauttilab.github.io/RCrashCourse/1_fundamentals_test.html) to test your knowledge.

Make sure you are familiar with this content as it will be needed for most of the lecture/tutorials.


***

# Topics Overview

 | Date | Topic
-----|--------|----------------------------------------------------
1 | 8-Jan | [Fundamental R](https://colauttilab.github.io/RCrashCourse/1_fundamentals_test.html) (self-study)
2 | 15-Jan | ggplot and professional visualizations in R
3 | 22-Jan | Data science and regular expressions (most people donâ€™t know what they are but they are super-useful!)
4 | 29-Jan | Reproducible R with Git and GitHub
5 | 5-Feb  | Introduction to Python
6 | 12-Feb | Introduction to Unix, Bash scripting, and the SLURM scheduler
- | 19-Feb | Reading Week
- | 26-Feb | No tutorial
7 | 4-Mar  | Bootstrap, randomization and permutation models
8 | 11-Mar | Simulations, functions and modelling
9 | 18-Mar | Principal components and factor analysis 
10 | 25-Mar | Cluster analysis
11 | 1-Apr |	Nonlinear regression
12 | 8-Apr  |	Sequence data and microbiome analysis in R


# Grading Rubric:

  * __10%__ Attendance and participation
  * __40%__ Assignments & quizzes
  * __50%__ Final project
    + __10%__ Proposal
    + __15%__ GitHub Code
    + __25%__ Poster
    
# Final Assignment

The goal of the final assignment is to apply the coding skills you have learned to address an interesting biological question, details will be posted in the coming weeks, but expect the following:

  1. Analyze a large dataset (>100,000 data points); original or published data are fine (e.g. geospatial data, RNA-seq/gene expression data)
  2. Write custom scripts using a combination of (1) Bash and (2) R or Python (or both)
  3. Produce informative and elegant visualizations
  4. All data manipulation, analysis and visualization hard-coded and archived on GitHub
  5. Project addresses an interesting biological question
  6. Create a poster for the project, and present it at the Biology graduate student symposium (April 20)

# Textbook:

No textbook is required for the course. I recommend Haddock & Dunn (H&D) - Practical Computing for Biologists, if you can get it at a reasonable price.
