---
title: |-
  BIOL 812 \
    Intro to Computation \
    and Big Data \
    in Biology
author: "Rob Colautti"
output: html_document
---

# Readings & Assignments {.tabset .tabset-fade .tabset-pills}   

## Jan 31

  1. Review __R Markdown__ and `ggplot2()` at the links below (see also link to ggplot2 Documentation noted in the qplot() tutorial)
  2. Use an R markdown file to create a clean, attractive 1-page report, focusing on a single visualization with ggplot(). The details of the report don’t matter (e.g. real vs. simulated data). 
  3. Use custom formatting to make the most attractive report you can make using R markdown and ggplot(). 
  4. You will be marked on:
      * aesthetic (clean & clear, with no unnecessary ‘ink’)
      * the application of methods used in Rmarkdown and ggplot()
      * clarity and readability of .Rmd code, and file knits without error.
  5. Submit __.Rmd__ file to [onq.queensu.ca](https://onq.queensu.ca) DropBox Link

## Jan 24  

Assignments due for R and Unix (see email from Jan 17)

## Jan 17    

Haddock & Dunn Chapters 4 & 5 (Windows users, use MobaXTerm)

  * Note: Okay to skip `curl` section if you are having problems

# Topics

## Computation Bootcamp:

  1. [R Crash Course](./RCrashCourse/1_fundamentals.html)
  2. [Introduction to the Shell, SLURM, bash and CAC](https://hpc-carpentry.github.io/hpc-intro/)
      + Unix/Linux [cheat sheet 1](https://files.fosswire.com/2007/08/fwunixref.pdf)
        + [cheat sheet 2](http://cheatsheetworld.com/programming/unix-linux-cheat-sheet/)
  3. Advanced R: 
      + [R Markdown](./RCrashCourse/5_advanced.html#2_r_markdown) 
      + see also:
          + Very handy .Rmd [cheat sheet](http://www.rstudio.com/wp-content/uploads/2016/03/rmarkdown-cheatsheet-2.0.pdf)
          + Detailed .Rmd [website](http://rmarkdown.rstudio.com/lesson-1.html)
          + [R Notebooks](http://rmarkdown.rstudio.com/r_notebooks.html#overview)
      + [Visualizations with `qplot()` and `ggplot()`](./RCrashCourse/2_qplot.html)
      + [Visualization lecture notes](./Lecures/Graphical_Concepts.pdf)
  4. Version Control with [Git & GitHub](http://swcarpentry.github.io/git-novice/)
  
### Coming soon...
  * [Regular Expressions (Regex) in R](./RCrashCourse/4_regex.html) 
          + General regex [cheat sheet](https://www.cheatography.com/davechild/cheat-sheets/regular-expressions/pdf/)
          + Regex [cheat sheet for R](https://www.rstudio.com/wp-content/uploads/2016/09/RegExCheatsheet.pdf)
      + [Custom functions in R](./RCrashCourse/5_advanced.html)
  * Python Crash Course
  * Bioconductor & Snakemake
  * Relational Databases
  
## Computation in Practice

  * Big Data in Ecology
  * Next Generation Sequencing

# Grading Rubric:

  * __10%__ Attendance and participation
  * __40%__ Assignments & quizzes
  * __50%__ Final project
    + __10%__ Proposal
    + __15%__ GitHub Code
    + __25%__ Poster
    
# Final Assignment

The goal of the final assignment is to apply the coding skills you have learned to address an interesting biological question, details will be posted in the coming weeks, but expect the following:

1. Analyze a large dataset (>1000 data points); original or published data are fine (e.g. geospatial data, RNA-seq/gene expression data)
2. Write custom scripts using a combination of (1) Bash and (2) R or Python (or both)
3. Produce informative and elegant visualizations
4. All data manipulation, analysis and visualization hard-coded and archived on GitHub
4. Present key results as a poster
5. Project addresses an interesting biological question


# Haddock & Dunn (H&D) - Pracical Computing for Biologists 

  * [Table of Contents]
  * [Chapter 1]
  * [Chapter 2]
  * [Chapter 3]
  * [Chapter 4]
  * [Chapter 5]
  * [Chapter 6]
  * [Chapter 7]
  * [Chapter 8]
  * [Chapter 9]
  * [Chapter 10]
  * [Chapter 11]
  * [Chapter 12]
  * [Chapter 13]
  * [Chapter 14]
  * [Chapter 15]
  * [Chapter 16]
  * [Chapter 17]
  * [Chapter 18]
  * [Chapter 19]
  * [Chapter 20]
  * [Chapter 21]
  * [Chapter 22]
  * [Appendix 1]
  * [Appendix 2]
  * [Appendix 3]

  





