---
title: "Colautti Lab Resources"
output: html_document
---

# COVID UPDATE

## IMPORTANT INFORMATION FOR ANYONE WORKING IN THE LAB: [Queen's Return to Campus Guidelines](https://www.queensu.ca/vpfa/sites/webpublish.queensu.ca.vpfawww/files/files/Return_to_Campus_Guidelines.pdf) 

# Essentials

## 1. Carefully read the [Colautti Lab Manifesto](./Manifesto/index.html)

  * Essential reading for all members of the Colautti Lab

## 2. Learn how to research

  * Take the online course from iBiology called "[Let's Experiment: A Guide for Scientists Working at the Bench](https://courses.ibiology.org/courses/course-v1:iBiology+LE+SP/about)". The course is focused on lab work but covers essential concepts in experimental biology.
  
## 3. Develop a project
  
  * Define your question and research the relevant theory
  * Identify your main hypotheses, generate quantitative predictions, and plan out your experiment(s).
  * Run your ideas by everyone who will listen, especially your mentor. Give it your best effort and then be prepared to change it.

## 4. Read the [BES Guides](https://www.britishecologicalsociety.org/publications/guides-to/) to Better Science

  * [Data Management](./Readings/BES-Data-Guide-2017.pdf)
  * [Reproducible Code](./Readings/BES-Reproducible-Code.pdf)
  * [Getting Published](./Readings/BES-Getting-Published-Guide-2017.pdf)
  * [Peer Review](./Readings/BES-Peer-Review-Guide-2017.pdf)
    * and a related [blog post](https://appliedecologistsblog.com/2019/09/18/quality-in-peer-review) on Peer Review

## 5. Collect Data

A good __rule of thumb:__ imagine somebody reading over your shoulder who is not familiar with your work. Would they be able to interpret your filenames, meta-data, README files, data field names, documentation, etc? A bonus is that you will be able to understand what you did when you revisit your data months or years later.

### BEFORE you start

**Organize and label samples carefully**

  * Any mistakes at this stage will propagate through your whole project and may never be detected!
  * We created the [baRcodeR package] in R to help with this. It's a simple as `install.packages(baRcodeR)`. Use it!
    * Quick Reference Guide: Choose _baRcodeR Cheatsheet_ from the _Addins_ menu in R Studio. Or download pdf from [FigShare](https://figshare.com/articles/baRcodeR_Cheat_Sheet/7043309).
    * CRAN website for[baRcodeR](https://cran.r-project.org/package=baRcodeR)
    * baRcodeR [Quick Start Guide](https://cran.r-project.org/web/packages/baRcodeR/vignettes/Using-baRcodeR.html) and [Graphical Interface Instruction](https://cran.r-project.org/web/packages/baRcodeR/vignettes/use-addin.html)
  * Read bioRxiv manuscript [**baRcodeR** with **PyTrackDat** Open-source labelling and tracking of biological samples for repeatable science](https://www.biorxiv.org/content/early/2018/10/30/457051) to get a better understanding of labelling and sample management.

**Create a Data Management Plan (DMP)**
  * Know this now -- you will have to publish your data publicly, so take a bit of extra time now to make sure it is understandable.
  * Using short-hand names (e.g. `Wt` instead of `total dry biomass`) will save a lot of coding headaches. 
  * Create a `metadata.txt` file describing the basics of your data. A short paragraph of where it comes from, and then a line describing each file and variable in your dataset.
  
### WHILE you collect (or simulate) data

**Collect and Manage Data CAREFULLY**

  * Build in redundancy. Never have your data in only 1 place.
  * Collect data by hand, using paper and a pencil. For larger experiments, print out data sheets with defined cells.
  * Enter handwritten data onto a computer and back it up online. Do it WHILE you record data, not after.
  * Manage data in one or more spreadsheets, or setup a database with [PyTrackDat](https://github.com/ColauttiLab/PyTrackDat/)
  * Don't worry about trying to put everything in one spreadsheet. As long as you have ID codes (e.g. from baRcodeR) to link different experiments/measurements then it is easy to connect these later in R.
  
### After Finishing

**Publish your data**

  * For data collected at QUBS, use the [QUBS Dataverse Repository](https://dataverse.scholarsportal.info/dataverse/QUBS), open to all
  * For other datasets, use a repository like [datadryad.org](https://colauttilab.github.io/), [FigShare.com](https://figshare.com/)
  * For larger collaborative projects, create a [Dataverse Repository](https://dataverse.scholarsportal.info/)
  * Use specialized datasets like [NCBI SRA database](https://www.ncbi.nlm.nih.gov/sra) or one on the list in [this paper](https://www.nature.com/sdata/policies/repositories), or search Google for a more appropriate home) 
  * Consider whether it would be useful to combine repositories e.g. upload SRA files to NCBI and include Accession numbers in metadata posted in another repository containing code and called SNPs.

## 7. Start Coding

__READ__ Google's [R Style Guide](https://google.github.io/styleguide/Rguide.xml)

__Set up__ R Studio ([Free Desktop Version](https://www.rstudio.com/products/rstudio/download/)) with a [GitHub account](https://support.rstudio.com/hc/en-us/articles/200532077-Version-Control-with-Git-and-SVN) on your computer.
    
### Develop Your Coding Superpowers

![](./Superheroes.jpg)

| __Conventional Superpower__ | __Coding Superpower__ |
|:---------------------------:|:---------------------:|
| Enhanced abilities          | Fundamentals of [R](./RCrashCourse/1_fundamentals.html) and [Python](./PythonCrashCourse/1_fundamentals.html) |
| Mind Control                | [Bash and Linux Command Line Computing](https://colauttilab.github.io/LinuxCrashCourse/Linux-Fundamentals.html) |
| Robot Army                  | Flow Control in [R](./RCrashCourse/1_fundamentals.html#7_flow_control) and [Python](./PythonCrashCourse/1_fundamentals.html#6_flow_control) |
| Clone backup                | [Git and GitHub](http://swcarpentry.github.io/git-novice/) |
| Illusionist                 | Custom Visualizations with [qplot](./RCrashCourse/2_qplot.html) and [ggplot](./RCrashCourse/3_ggplot.html) |
| Telekenesis                 | Custom Functions in [R](./RCrashCourse/5_advanced.html) and [Python](PythonCrashCourse/1_fundamentals.html#7_custom_functions) |
| Transmutation               | Regular Expresssions in [R](RCrashCourse/4_regex.html) and [Python](./PythonCrashCourse/2_regex.html) |
| Mega-mind                   | Resampling & Simulations in R [Pt.1](./EcologyTutorials/bootstrap.html) [Pt.2](./EcologyTutorials/simulations.html) [Pt.3](ColauttiLab.github.io/EcologyTutorials/maxLik_MCMC.html) |
| Master Builder              | Custom [R Packages](./RCrashCourse/Package_tutorial.html) |


## 7. More coding

  * [Statistical Analysis in R](./RintroStats/1_Distributions.html)
  * [Population Genetics in R](./PopGen/1_Introduction.html)
  * [Maximum Liklihood & MCMC](./EcologyTutorials/maxLik_MCMC.html)

### Ecology

  * [Geographical maps in R](./EcologyTutorials/mapping.html)
  * [GIS and Climate data in R](./EcologyTutorials/GGMFS_climate.html)

### Genomics

  * [_de novo_ Genome Assembly](./NGS/deNovoTutorial.html)
  * [RNA alignment and Transcriptome Analysis](./NGS/TuxedoTutorial.html)

***

# Course material

## Colautti's courses

[2018 HPC Summer School @ Queens](HPC2018.html)

[2018 SETAC Short-Course](SETAC.html)

[BIOL 307](FieldMethods.html) -- Field Methods in Ecological and Environmental Genomics

[BIOL 432](biol432.html) -- Introduction to Computation and Big Data in Biology

[BIOL 812](biol812.html) -- Applied Computation for Biologists

## Other sourse material

__Gram Coop's PopGen notes__

  * [Release Notes](https://github.com/cooplab/popgen-notes/blob/master/release_popgen_notes.pdf)
  * [Main GitHub Site](https://github.com/cooplab/popgen-notes)
  
***

# Other resources

__Public Website__

* [bit.ly/colautti](http://bit.ly/colautti) -- Typical departmental website with basic research information

__Publication Record__

* [Google Scholar Website](http://scholar.google.ca/citations?user=J7MvakAAAAAJ&hl=en)
* [ORCiD](https://orcid.org/0000-0003-4213-0711)
* [ResearcherID](http://www.researcherid.com/rid/E-6804-2011)

__Public Engagement__

* Follow @ColauttiLab on [Twitter](https://twitter.com/ColauttiLab)

__Private Lab Website__

* [ecogeno.wordpress.com](https://ecogeno.wordpress.com) -- Lab protocols, analytical methods, etc.



