<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>R Crash Course for Biologists, 2nd Edition</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./favicon.ico" rel="icon">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-HZY3MS2C2R"></script>

<script type="text/javascript">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-HZY3MS2C2R', { 'anonymize_ip': true});
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="https://EcoEvoGeno.org" class="navbar-brand navbar-brand-logo">
    <img src="./ColauttiLabLogo.png" alt="Colautti Lab logo" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="https://EcoEvoGeno.org">
    <span class="navbar-title">R Crash Course for Biologists, 2nd Edition</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-introduction" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Introduction</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-introduction">    
        <li>
    <a class="dropdown-item" href="./index.html">
 <span class="dropdown-text">Book Info</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./preface.html">
 <span class="dropdown-text">Preface</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-basics" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Basics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-basics">    
        <li>
    <a class="dropdown-item" href="./fundamentals.html">
 <span class="dropdown-text">R Fundamentals</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./visualizations.html">
 <span class="dropdown-text">Quick Visualizations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./rmarkdown.html">
 <span class="dropdown-text">R Markdown</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-more-graphs" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">More Graphs</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-more-graphs">    
        <li>
    <a class="dropdown-item" href="./customizations.html">
 <span class="dropdown-text">Basic Customization</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./advanced_vis.html">
 <span class="dropdown-text">Advanced Visualizations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./multiplot.html">
 <span class="dropdown-text">Multi-plot Graphs</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-advanced" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Advanced</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-advanced">    
        <li>
    <a class="dropdown-item" href="./datascience.html">
 <span class="dropdown-text">Data Science</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./flowcontrol.html">
 <span class="dropdown-text">Flow Control</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./regex.html">
 <span class="dropdown-text">Regular Expressions I</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./regex2.html">
 <span class="dropdown-text">Regular Expressions II</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./customfunctions.html">
 <span class="dropdown-text">Custom Functions</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./matrices.html">
 <span class="dropdown-text">Matrices</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="./conclusion.html"> 
<span class="menu-text">Conclusion</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#data-science" id="toc-data-science" class="nav-link active" data-scroll-target="#data-science">Data Science</a>
  <ul class="collapse">
  <li><a href="#introduction" id="toc-introduction" class="nav-link" data-scroll-target="#introduction">Introduction</a>
  <ul class="collapse">
  <li><a href="#data-science-life-cycle" id="toc-data-science-life-cycle" class="nav-link" data-scroll-target="#data-science-life-cycle">Data Science Life-Cycle</a></li>
  </ul></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#d-data-wrangling" id="toc-d-data-wrangling" class="nav-link" data-scroll-target="#d-data-wrangling">2D Data Wrangling</a>
  <ul class="collapse">
  <li><a href="#tibbles-and-readr" id="toc-tibbles-and-readr" class="nav-link" data-scroll-target="#tibbles-and-readr"><code>tibbles</code> and <code>readr()</code></a></li>
  <li><a href="#filter-rows" id="toc-filter-rows" class="nav-link" data-scroll-target="#filter-rows"><code>filter()</code> Rows</a></li>
  <li><a href="#rename-columns" id="toc-rename-columns" class="nav-link" data-scroll-target="#rename-columns"><code>rename()</code> Columns</a></li>
  <li><a href="#arrange-rows" id="toc-arrange-rows" class="nav-link" data-scroll-target="#arrange-rows"><code>arrange()</code> Rows</a></li>
  <li><a href="#select-columns" id="toc-select-columns" class="nav-link" data-scroll-target="#select-columns"><code>select()</code> Columns</a></li>
  <li><a href="#everything" id="toc-everything" class="nav-link" data-scroll-target="#everything"><code>everything()</code></a></li>
  <li><a href="#mutate-columns" id="toc-mutate-columns" class="nav-link" data-scroll-target="#mutate-columns"><code>mutate()</code> Columns</a></li>
  <li><a href="#transmute-columns" id="toc-transmute-columns" class="nav-link" data-scroll-target="#transmute-columns"><code>transmute()</code> Columns</a></li>
  <li><a href="#summarise-group_by" id="toc-summarise-group_by" class="nav-link" data-scroll-target="#summarise-group_by"><code>summarise()</code> + <code>group_by()</code></a></li>
  <li><a href="#weighted-mean" id="toc-weighted-mean" class="nav-link" data-scroll-target="#weighted-mean">Weighted Mean</a></li>
  <li><a href="#pipe" id="toc-pipe" class="nav-link" data-scroll-target="#pipe"><code>%&gt;%</code> (pipe)</a></li>
  </ul></li>
  <li><a href="#join-datasets" id="toc-join-datasets" class="nav-link" data-scroll-target="#join-datasets">Join Datasets</a></li>
  <li><a href="#wide-vs-long-data" id="toc-wide-vs-long-data" class="nav-link" data-scroll-target="#wide-vs-long-data"><em>Wide</em> vs <em>Long</em> Data</a>
  <ul class="collapse">
  <li><a href="#pivot_longer" id="toc-pivot_longer" class="nav-link" data-scroll-target="#pivot_longer"><code>pivot_longer</code></a></li>
  <li><a href="#pivot_wider" id="toc-pivot_wider" class="nav-link" data-scroll-target="#pivot_wider"><code>pivot_wider</code></a></li>
  </ul></li>
  <li><a href="#missing-data" id="toc-missing-data" class="nav-link" data-scroll-target="#missing-data">Missing Data</a>
  <ul class="collapse">
  <li><a href="#na-and-na.rm" id="toc-na-and-na.rm" class="nav-link" data-scroll-target="#na-and-na.rm"><code>NA</code> and <code>na.rm()</code></a></li>
  <li><a href="#na-vs-0" id="toc-na-vs-0" class="nav-link" data-scroll-target="#na-vs-0"><code>NA</code> vs <code>0</code></a></li>
  <li><a href="#is.na" id="toc-is.na" class="nav-link" data-scroll-target="#is.na"><code>is.na()</code></a></li>
  </ul></li>
  <li><a href="#naughty-data" id="toc-naughty-data" class="nav-link" data-scroll-target="#naughty-data">Naughty Data</a></li>
  <li><a href="#dates" id="toc-dates" class="nav-link" data-scroll-target="#dates">Dates</a>
  <ul class="collapse">
  <li><a href="#run-time" id="toc-run-time" class="nav-link" data-scroll-target="#run-time">Run Time</a></li>
  <li><a href="#date-objects" id="toc-date-objects" class="nav-link" data-scroll-target="#date-objects"><code>date</code> Objects</a></li>
  <li><a href="#datetime" id="toc-datetime" class="nav-link" data-scroll-target="#datetime"><code>datetime</code></a></li>
  <li><a href="#extracting-components" id="toc-extracting-components" class="nav-link" data-scroll-target="#extracting-components">Extracting Components</a></li>
  <li><a href="#categorical-dates" id="toc-categorical-dates" class="nav-link" data-scroll-target="#categorical-dates">Categorical Dates</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">




<section id="data-science" class="level1">
<h1>Data Science</h1>
<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Data Science is a relatively new field of study that merges <strong>computer science</strong> and <strong>statistics</strong> to answer questions in other domains (e.g.&nbsp;business, medicine, biology, psychology). Data Science as a discipline has grown in popularity in response to the rapid rate of increase in data collection and publication.</p>
<p>Data Science often involves ‘Big Data’, which doesn’t have a strict quantitative definition but will usually have one or more of the following characteristics:</p>
<ol type="1">
<li>High <strong>Volume</strong> – large file sizes with may observations.</li>
<li>Wide <strong>Variety</strong> – many different types of data.</li>
<li>High <strong>Velocity</strong> – data accumulates at a high rate.</li>
<li>Compromised <strong>Veracity</strong> – data quality issues must be addressed otherwise downstream analyses will be compromised.</li>
</ol>
<blockquote class="blockquote">
<p><strong>Question</strong>: What are some examples of ‘big data’ in Biology?</p>
</blockquote>
<p><strong>Answer</strong>: There are many types of biological data that could be listed, but medical records, remote sensing data, and ‘omics data are common examples of ’big data’ in biology.</p>
<p>In biology, it can be helpful to think of Data Science as a continuous life-cycle with multiple stages:</p>
<section id="data-science-life-cycle" class="level3">
<h3 class="anchored" data-anchor-id="data-science-life-cycle">Data Science Life-Cycle</h3>
<ol type="1">
<li><p><strong>Hypothesize</strong> – Make initial observations about the natural world, or insights from other data, that lead to testable hypotheses. Your core biology training is crucial here.</p></li>
<li><p><strong>Collect</strong> – This may involve taking measurements yourself, manually entering data that is not yet in electronic format, requesting data from authors of published studies, or importing data from online sources. Collecting data is a crucial step that is often done poorly.</p></li>
<li><p><strong>Correct</strong> – Investigate the data for quality assurance, to identify and fix potential errors. Start to visualize the data to look for outliers, odd frequency distributions, or nonsensical relationships among variables.</p></li>
<li><p><strong>Explore</strong> – Try to understand the data, where they come from, and potential limitations on their use. Continue visualizing data; this may cause you to modify your hypotheses slightly.</p></li>
<li><p><strong>Model</strong> – Now that hypotheses are clearly defined, apply statistical tests of their validity.</p></li>
<li><p><strong>Report</strong> – Use visualizations along with the results of your statistical tests to summarise your findings.</p></li>
<li><p><strong>Repeat</strong> – Return to step 1.</p></li>
</ol>
<p>In this chapter, we focus mainly on coding in R for steps 2, 3, and 6. Step 5 requires a firm understanding of statistics, which is the focus of the book <em>R STATS Crash Course for Biologists</em>. Visualizations in Steps 3 &amp; 4 were covered in earlier chapters.. Step 1 requires a good understanding of the study system, as covered in a typical university degree in the biological sciences.</p>
<p>Data collection and management are crucial steps in the Data Science Life-Cycle. Read the <code>baRcodeR</code> paper by Wu et al (2022) (<a href="https://doi.org/10.1111/2041-210X.13405" class="uri"><em>https://doi.org/10.1111/2041-210X.13405</em></a>). called <em>baRcodeR with PyTrackDat: Open-source labelling and tracking of biological samples for repeatable science.</em> Pay particular attention to the ‘<em>Data Standards</em>’ section. The <em>baRcodeR</em> and <em>PyTrackDat</em> programs and their application to current projects may also be of interest.</p>
</section>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<p>The <strong>tidyverse</strong> library is a set of packages created by the same developers responsible for R Studio. There are a number of packages and functions that improve on base R. The name is based on the idea of living in a data universe that is neat and <em>tidy</em>.</p>
<p>The book <em>R for Data Science</em> by Hadley Wickham &amp; Garrett Grolemund is an excellent resource for learning about the tidyverse. In this chapter we’ll touch on a few of the main packages and functions.</p>
<blockquote class="blockquote">
<p><strong>Protip</strong> In general, any book by Hadley Wickham that you come across is worth reading if you want to be proficient in R.</p>
</blockquote>
<p>While we wait for the <code>tidyverse</code> to install, let’s consider a few general principles of data science.</p>
</section>
<section id="d-data-wrangling" class="level2">
<h2 class="anchored" data-anchor-id="d-data-wrangling">2D Data Wrangling</h2>
<p>The <code>dplyr</code> library in R has many useful features for importing and reorganizing your data for steps 2, 3 and 4 in the Data Science Life-Cycle.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'dplyr' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Note</strong>: This error message informs us that the <code>dplyr</code> package uses function or parameter names that are the same as other base or stats packages in R. These base/stats functions are ‘masked’ meaning that when you run one (e.g.&nbsp;<code>filter</code>) then R will run the <code>dplyr</code> version rather than the stats version.</p>
</blockquote>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We’ll work with our <code>FallopiaData.csv</code> data set, and remind ourselves of the structure of the data</p>
<section id="tibbles-and-readr" class="level3">
<h3 class="anchored" data-anchor-id="tibbles-and-readr"><code>tibbles</code> and <code>readr()</code></h3>
<p>We looked at <code>data.frame</code> objects in the first chapter as an expansion of <em>matrices</em> with a few additional features like column and row names. A <code>tibble</code> is the <code>tidyverse</code> version of the <code>data.frame</code> object and includes a few more useful features. To import a dataset to a <code>tibble</code> instead of a <code>data.frame</code> object, we use <code>read_csv</code> instead of <code>read.csv</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyverse' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'forcats' was built under R version 4.3.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'lubridate' was built under R version 4.3.2</code></pre>
</div>
</div>
<p>You may also see a message about conflicts here, with the <code>filter()</code> and <code>lag()</code> functions from <code>dplyr</code> masking the functions from the <code>stats</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>Fallo<span class="ot">&lt;-</span><span class="fu">read_csv</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"https://colauttilab.github.io/RCrashCourse/FallopiaData.csv"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(Fallo)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 123 Columns: 13
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Scenario, Nutrients, Taxon
dbl (10): PotNum, Symphytum, Silene, Urtica, Geranium, Geum, All_Natives, Fa...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [123 × 13] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ PotNum      : num [1:123] 1 2 3 5 6 7 8 9 10 11 ...
 $ Scenario    : chr [1:123] "low" "low" "low" "low" ...
 $ Nutrients   : chr [1:123] "low" "low" "low" "low" ...
 $ Taxon       : chr [1:123] "japon" "japon" "japon" "japon" ...
 $ Symphytum   : num [1:123] 9.81 8.64 2.65 1.44 9.15 ...
 $ Silene      : num [1:123] 36.4 29.6 36 21.4 23.9 ...
 $ Urtica      : num [1:123] 16.08 5.59 17.09 12.39 5.19 ...
 $ Geranium    : num [1:123] 4.68 5.75 5.13 5.37 0 9.05 3.51 9.64 7.3 6.36 ...
 $ Geum        : num [1:123] 0.12 0.55 0.09 0.31 0.17 0.97 0.4 0.01 0.47 0.33 ...
 $ All_Natives : num [1:123] 67 50.2 61 40.9 38.4 ...
 $ Fallopia    : num [1:123] 0.01 0.04 0.09 0.77 3.4 0.54 2.05 0.26 0 0 ...
 $ Total       : num [1:123] 67.1 50.2 61.1 41.7 41.8 ...
 $ Pct_Fallopia: num [1:123] 0.01 0.08 0.15 1.85 8.13 1.12 3.7 0.61 0 0 ...
 - attr(*, "spec")=
  .. cols(
  ..   PotNum = col_double(),
  ..   Scenario = col_character(),
  ..   Nutrients = col_character(),
  ..   Taxon = col_character(),
  ..   Symphytum = col_double(),
  ..   Silene = col_double(),
  ..   Urtica = col_double(),
  ..   Geranium = col_double(),
  ..   Geum = col_double(),
  ..   All_Natives = col_double(),
  ..   Fallopia = col_double(),
  ..   Total = col_double(),
  ..   Pct_Fallopia = col_double()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
</div>
<p>This file is an example of a 2-dimensional data set, which is common in biology. 2D datasets have the familiar row x column layout used by spreadsheet programs like Microsoft Excel or Google Sheets. There are some exceptions, but data in this format should typically follows 3 rules:</p>
<ol type="1">
<li>Each cell contains a single value</li>
<li>Each variable must have its own column</li>
<li>Each observation must have its own row</li>
</ol>
<p>Making sure your data are arranged this way will usually make it much easier to work with.</p>
</section>
<section id="filter-rows" class="level3">
<h3 class="anchored" data-anchor-id="filter-rows"><code>filter()</code> Rows</h3>
<p>The <code>filter()</code> function will subset observations based on values of interest within particular columns of our dataset. For example, we may want to <em>filter</em> the rows (i.e.&nbsp;pots) that had at least 70 <span class="math inline">\(g\)</span> of total biomass.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>Pot1<span class="ot">&lt;-</span><span class="fu">filter</span>(Fallo,Total <span class="sc">&gt;=</span> <span class="dv">70</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Pot1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
  PotNum Scenario Nutrients Taxon Symphytum Silene Urtica Geranium  Geum
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;
1     60 high     high      bohem      7.77   51.4   5.13    10.1   0.37
2     67 gradual  high      japon      2.92   25.2  19.1     23.1   0.06
3     70 gradual  high      japon     10.1    47.0  18.6      0.64  0.82
4     86 gradual  high      bohem      2.93   60.9   4.11     6.67  1.27
5     95 extreme  high      japon      4.92   25.9  40.3      4.92  0.07
6    103 extreme  high      japon      6.92   49.4   0       10.3   0.42
# ℹ 4 more variables: All_Natives &lt;dbl&gt;, Fallopia &lt;dbl&gt;, Total &lt;dbl&gt;,
#   Pct_Fallopia &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="rename-columns" class="level3">
<h3 class="anchored" data-anchor-id="rename-columns"><code>rename()</code> Columns</h3>
<p>There are different options to change the names of columns in your data. In base R you can use the <code>names()</code> function with the square bracket index <code>[]</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span>Fallo</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "PotNum"       "Scenario"     "Nutrients"    "Taxon"        "Symphytum"   
 [6] "Silene"       "Urtica"       "Geranium"     "Geum"         "All_Natives" 
[11] "Fallopia"     "Total"        "Pct_Fallopia"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(X)[<span class="dv">12</span>]<span class="ot">&lt;-</span><span class="st">"Total_Biomass"</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "PotNum"        "Scenario"      "Nutrients"     "Taxon"        
 [5] "Symphytum"     "Silene"        "Urtica"        "Geranium"     
 [9] "Geum"          "All_Natives"   "Fallopia"      "Total_Biomass"
[13] "Pct_Fallopia" </code></pre>
</div>
</div>
<p>There is also a simple <code>dplyr</code> function to do this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">rename</span>(Fallo, <span class="at">Total_Biomass =</span> Total)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "PotNum"        "Scenario"      "Nutrients"     "Taxon"        
 [5] "Symphytum"     "Silene"        "Urtica"        "Geranium"     
 [9] "Geum"          "All_Natives"   "Fallopia"      "Total_Biomass"
[13] "Pct_Fallopia" </code></pre>
</div>
</div>
</section>
<section id="arrange-rows" class="level3">
<h3 class="anchored" data-anchor-id="arrange-rows"><code>arrange()</code> Rows</h3>
<p>Use the <code>arrange()</code> function to sort the <em>rows</em> of your data based on the values in one or more <em>columns</em>. For example, let’s re-arrange our FallopiaData.csv dataset based on Taxon (a string denoting the species of Fallopia used) and Total (a float denoting the total biomass in each pot).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">arrange</span>(Fallo, Taxon, Total)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
  PotNum Scenario Nutrients Taxon Symphytum Silene Urtica Geranium  Geum
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;
1     26 low      low       bohem     13.2    18.1   0        0     0.1 
2     17 low      low       bohem      4.9    29.5   1.36     0     0.19
3     80 gradual  high      bohem     11.9    17.2   8.92     0.94  0.18
4     18 low      low       bohem      3.51   27.6   8.14     3.81  0.21
5     28 low      low       bohem     10.6    18.8   7.19     6.73  0.17
6     22 low      low       bohem      0.76   22.7   9.85    10.6   0.74
# ℹ 4 more variables: All_Natives &lt;dbl&gt;, Fallopia &lt;dbl&gt;, Total &lt;dbl&gt;,
#   Pct_Fallopia &lt;dbl&gt;</code></pre>
</div>
</div>
<p>use the <code>desc()</code> function with <code>arrange()</code> to reverse and sort in <em>descending</em> order. We can sort by multiple columns in the <code>by_group=</code> parameter.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">arrange</span>(Fallo, <span class="at">by_group=</span>Taxon, <span class="fu">desc</span>(Silene))</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
  PotNum Scenario Nutrients Taxon Symphytum Silene Urtica Geranium  Geum
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;
1     86 gradual  high      bohem      2.93   60.9   4.11     6.67  1.27
2     53 high     high      bohem      7.05   56.3   1.14     4.07  0   
3     60 high     high      bohem      7.77   51.4   5.13    10.1   0.37
4     50 high     high      bohem      8.52   44.6   1.27     9.45  0.35
5     79 gradual  high      bohem     11.0    44.6   1.56     0.03  0.05
6    107 extreme  high      bohem      0      43.8   8.1      8.18  0.36
# ℹ 4 more variables: All_Natives &lt;dbl&gt;, Fallopia &lt;dbl&gt;, Total &lt;dbl&gt;,
#   Pct_Fallopia &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="select-columns" class="level3">
<h3 class="anchored" data-anchor-id="select-columns"><code>select()</code> Columns</h3>
<p>The <code>select()</code> function can be used to select a subset of columns (i.e.&nbsp;variables) from your data.</p>
<p>Suppose we only want to look at total biomass, but keep all the treatment columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, PotNum, Scenario, Nutrients, Taxon, Total)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  PotNum Scenario Nutrients Taxon Total
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;dbl&gt;
1      1 low      low       japon  67.1
2      2 low      low       japon  50.2
3      3 low      low       japon  61.1
4      5 low      low       japon  41.7
5      6 low      low       japon  41.8
6      7 low      low       japon  48.3</code></pre>
</div>
</div>
<p>You can also use the colon <code>:</code> to select a range of columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, PotNum<span class="sc">:</span>Taxon, Total)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  PotNum Scenario Nutrients Taxon Total
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt; &lt;dbl&gt;
1      1 low      low       japon  67.1
2      2 low      low       japon  50.2
3      3 low      low       japon  61.1
4      5 low      low       japon  41.7
5      6 low      low       japon  41.8
6      7 low      low       japon  48.3</code></pre>
</div>
</div>
<p>Exclude columns with <code>-</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, <span class="sc">-</span>PotNum<span class="sc">:</span>Taxon, <span class="sc">-</span>Total)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in x:y: numerical expression has 12 elements: only the first used</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Oops, what generated that warning? Take a careful look at the error message and see if you can figure it out.</p>
</blockquote>
<p>The problem is we are using the range of columns between PotNum and Taxon, but in one case we are excluding and the other we are including. We need to be consistent:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, <span class="sc">-</span>PotNum<span class="sc">:-</span>Taxon, Total)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  Symphytum Silene Urtica Geranium  Geum All_Natives Fallopia Total Pct_Fallopia
      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;
1      9.81   36.4  16.1      4.68  0.12        67.0     0.01  67.1         0.01
2      8.64   29.6   5.59     5.75  0.55        50.2     0.04  50.2         0.08
3      2.65   36.0  17.1      5.13  0.09        61.0     0.09  61.1         0.15
4      1.44   21.4  12.4      5.37  0.31        40.9     0.77  41.7         1.85
5      9.15   23.9   5.19     0     0.17        38.4     3.4   41.8         8.13
6      6.31   24.4   7        9.05  0.97        47.7     0.54  48.3         1.12</code></pre>
</div>
</div>
<p>Or a bit more clear:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, <span class="sc">-</span>(PotNum<span class="sc">:</span>Taxon), Scenario)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 10
  Symphytum Silene Urtica Geranium  Geum All_Natives Fallopia Total Pct_Fallopia
      &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;       &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;        &lt;dbl&gt;
1      9.81   36.4  16.1      4.68  0.12        67.0     0.01  67.1         0.01
2      8.64   29.6   5.59     5.75  0.55        50.2     0.04  50.2         0.08
3      2.65   36.0  17.1      5.13  0.09        61.0     0.09  61.1         0.15
4      1.44   21.4  12.4      5.37  0.31        40.9     0.77  41.7         1.85
5      9.15   23.9   5.19     0     0.17        38.4     3.4   41.8         8.13
6      6.31   24.4   7        9.05  0.97        47.7     0.54  48.3         1.12
# ℹ 1 more variable: Scenario &lt;chr&gt;</code></pre>
</div>
</div>
</section>
<section id="everything" class="level3">
<h3 class="anchored" data-anchor-id="everything"><code>everything()</code></h3>
<p>Use the <code>everything()</code> function with <code>select()</code> to rearrange your columns without losing any:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">select</span>(Fallo, Taxon, Scenario, Nutrients, PotNum, </span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>          Pct_Fallopia, <span class="fu">everything</span>())</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 13
  Taxon Scenario Nutrients PotNum Pct_Fallopia Symphytum Silene Urtica Geranium
  &lt;chr&gt; &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt;        &lt;dbl&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;
1 japon low      low            1         0.01      9.81   36.4  16.1      4.68
2 japon low      low            2         0.08      8.64   29.6   5.59     5.75
3 japon low      low            3         0.15      2.65   36.0  17.1      5.13
4 japon low      low            5         1.85      1.44   21.4  12.4      5.37
5 japon low      low            6         8.13      9.15   23.9   5.19     0   
6 japon low      low            7         1.12      6.31   24.4   7        9.05
# ℹ 4 more variables: Geum &lt;dbl&gt;, All_Natives &lt;dbl&gt;, Fallopia &lt;dbl&gt;,
#   Total &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="mutate-columns" class="level3">
<h3 class="anchored" data-anchor-id="mutate-columns"><code>mutate()</code> Columns</h3>
<p>Suppose we want to make a new column to our data frame or tibble. For example, we calculate the sum of biomass of <code>Urtica</code> and <code>Geranium</code> only. In base R, we could use <code>$</code> to select the column from the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span>Fallo</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>X<span class="sc">$</span>UrtSil<span class="ot">&lt;-</span>X<span class="sc">$</span>Urtica<span class="sc">+</span>X<span class="sc">$</span>Silene</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the <code>dplyr</code> package we can use the <code>mutate()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">mutate</span>(Fallo, <span class="at">UrtSil =</span> Urtica <span class="sc">+</span> Silene)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 14
  PotNum Scenario Nutrients Taxon Symphytum Silene Urtica Geranium  Geum
   &lt;dbl&gt; &lt;chr&gt;    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;
1      1 low      low       japon      9.81   36.4  16.1      4.68  0.12
2      2 low      low       japon      8.64   29.6   5.59     5.75  0.55
3      3 low      low       japon      2.65   36.0  17.1      5.13  0.09
4      5 low      low       japon      1.44   21.4  12.4      5.37  0.31
5      6 low      low       japon      9.15   23.9   5.19     0     0.17
6      7 low      low       japon      6.31   24.4   7        9.05  0.97
# ℹ 5 more variables: All_Natives &lt;dbl&gt;, Fallopia &lt;dbl&gt;, Total &lt;dbl&gt;,
#   Pct_Fallopia &lt;dbl&gt;, UrtSil &lt;dbl&gt;</code></pre>
</div>
</div>
<p>This is a lot more readable, especially when you have complicated equations or you want to add many of new columns.</p>
<blockquote class="blockquote">
<p><strong>Question</strong>: What if you only wanted to retain the new columns and delete everything else? Try it.</p>
</blockquote>
<p>Which functions did you use?</p>
</section>
<section id="transmute-columns" class="level3">
<h3 class="anchored" data-anchor-id="transmute-columns"><code>transmute()</code> Columns</h3>
<p>The <code>transmute()</code> functions acts as a combination of <code>mutate()</code> + <code>select()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">transmute</span>(Fallo, <span class="at">UrtSil =</span> Urtica <span class="sc">+</span> Silene)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(X)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 1
  UrtSil
   &lt;dbl&gt;
1   52.4
2   35.2
3   53.1
4   33.8
5   29.1
6   31.4</code></pre>
</div>
</div>
</section>
<section id="summarise-group_by" class="level3">
<h3 class="anchored" data-anchor-id="summarise-group_by"><code>summarise()</code> + <code>group_by()</code></h3>
<p>This can be useful for quickly summarizing your data, for example to find the mean or standard deviation based on a particular treatment or group.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>TrtGrp<span class="ot">&lt;-</span><span class="fu">group_by</span>(Fallo,Taxon,Scenario,Nutrients)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise</span>(TrtGrp, <span class="at">Mean=</span><span class="fu">mean</span>(Total), <span class="at">SD=</span><span class="fu">sd</span>(Total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'Taxon', 'Scenario'. You can override using
the `.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 5
# Groups:   Taxon, Scenario [10]
   Taxon Scenario     Nutrients  Mean    SD
   &lt;chr&gt; &lt;chr&gt;        &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;
 1 bohem extreme      high       58.3  7.34
 2 bohem fluctuations high       58.4  9.20
 3 bohem gradual      high       57.5  9.34
 4 bohem high         high       60.3  8.68
 5 bohem low          low        48.0  8.86
 6 japon extreme      high       57.2 10.9 
 7 japon fluctuations high       56.4 13.7 
 8 japon gradual      high       59.7  9.57
 9 japon high         high       56.4  8.20
10 japon low          low        52.0  8.29</code></pre>
</div>
</div>
</section>
<section id="weighted-mean" class="level3">
<h3 class="anchored" data-anchor-id="weighted-mean">Weighted Mean</h3>
<p>In our dataset, the <strong>Taxon</strong> column shows which of two species of <em>Fallopia</em> were used in the competition experiments. We might want to take the mean total biomass for each of the two <em>Fallopia</em> species.</p>
<p>However, there are other factors in our experiment that may affect biomass. The <em>Nutrients</em> column tells us whether pots received high or low nutrients, and this also affects biomass:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">group_by</span>(Fallo,Nutrients)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise</span>(X, <span class="at">Mean=</span><span class="fu">mean</span>(Total), <span class="at">SD=</span><span class="fu">sd</span>(Total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  Nutrients  Mean    SD
  &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt;
1 high       58.0  9.61
2 low        49.9  8.66</code></pre>
</div>
</div>
<p>We can see that the <code>Nutrients</code> treatment had a large effect on the mean biomass of a pot. Now imagine if our sampling design is ‘unbalanced’. For example, maybe we had some plant mortality or lost some plants to a tornado. If one of the two species in the <code>Taxon</code> column had more high-nutrient pots left over, then it would have a higher mean. BUT, what if the nutrients promoted growth? We would expect to see a difference in the mean of each <code>Taxon</code> group, this expected difference is just an artifact of unbalanced replication of the high nutrient treatment. We can simulate this effect by re-shuffling the species names:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>RFallo<span class="ot">&lt;-</span>Fallo</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">256</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>RFallo<span class="sc">$</span>Taxon<span class="ot">&lt;-</span><span class="fu">rbinom</span>(<span class="fu">nrow</span>(RFallo),<span class="at">size=</span><span class="dv">1</span>,<span class="at">prob=</span><span class="fl">0.7</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">group_by</span>(RFallo,Taxon)</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise</span>(X, <span class="at">Mean=</span><span class="fu">mean</span>(Total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  Taxon  Mean
  &lt;int&gt; &lt;dbl&gt;
1     0  56.1
2     1  56.5</code></pre>
</div>
</div>
<p>In this example, the difference is less than $ 1%$ of the mean (i.e., <span class="math inline">\(0.4 / 56.1 = 0.0071 = 0.71%\)</span>), but this could be due to the imbalance offsetting biological differences between the taxa. In other scenarios it could be much larger. To fix this problem, we may want to take a <strong>weighted mean</strong>. In this case, we want to <em>weight</em> the mean of each taxon by the unbalanced <code>Scenario</code> and <code>Treatment</code> categories.</p>
<p><strong>Step 1</strong>: Group by all three columns and calculate group means</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>X1<span class="ot">&lt;-</span><span class="fu">group_by</span>(RFallo,Taxon,Scenario,Nutrients)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>(Y1<span class="ot">&lt;-</span><span class="fu">summarise</span>(X1,<span class="at">Mean=</span><span class="fu">mean</span>(Total)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
# Groups:   Taxon, Scenario [10]
   Taxon Scenario     Nutrients  Mean
   &lt;int&gt; &lt;chr&gt;        &lt;chr&gt;     &lt;dbl&gt;
 1     0 extreme      high       54.7
 2     0 fluctuations high       58.4
 3     0 gradual      high       56.2
 4     0 high         high       59.8
 5     0 low          low        50.2
 6     1 extreme      high       59.1
 7     1 fluctuations high       56.8
 8     1 gradual      high       59.9
 9     1 high         high       57.6
10     1 low          low        49.8</code></pre>
</div>
</div>
<p>Now we have a single value for each of the subgroups. Instead of over-representing pots from <code>Taxon=1</code>, we have equal weightings of all three groups.</p>
<p><strong>Step 2</strong>: Take the output from Step 1 and calculate the <em>mean of means</em> for the column of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>X2<span class="ot">&lt;-</span><span class="fu">group_by</span>(Y1,Taxon)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>Y2<span class="ot">&lt;-</span><span class="fu">summarise</span>(X2, <span class="at">Mean=</span><span class="fu">mean</span>(Mean))</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(Y2,<span class="fu">desc</span>(Mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  Taxon  Mean
  &lt;int&gt; &lt;dbl&gt;
1     1  56.6
2     0  55.8</code></pre>
</div>
</div>
<p>Now there is more than a <span class="math inline">\(2.1 %\)</span> difference (i.e., <span class="math inline">\(1.2 / 55.8 = 0.0215\)</span>) between the taxa. This is triple the effect that we calculated with unbalanced samples!</p>
</section>
<section id="pipe" class="level3">
<h3 class="anchored" data-anchor-id="pipe"><code>%&gt;%</code> (pipe)</h3>
<p>The <code>dplyr</code> package includes a special command (<code>%&gt;%</code>) called a <em>pipe</em>. The <em>pipe</em> is a very handy tool for complex data wrangling. It allows us to string together multiple functions and then <em>pipe</em> our data from one function to the next. In principle, this is similar to the multi-function plotting commands we made with <code>ggplot()</code> in earlier chapters.</p>
<p>The pipe is useful to combine operations without creating a whole bunch of new objects. This can save on memory use and run speed because you are not making new objects for every single function.</p>
<blockquote class="blockquote">
<p><strong>Pro-tip</strong>: Use <code>Ctl</code> + <code>Shift</code> + <code>m</code> to add a pipe quickly</p>
</blockquote>
<p>For example, we can re-write the weighted mean example using pipes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>RFallo <span class="sc">%&gt;%</span> </span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Taxon,Scenario,Nutrients) <span class="sc">%&gt;%</span> </span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Mean=</span><span class="fu">mean</span>(Total)) <span class="sc">%&gt;%</span> </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Taxon,Scenario) <span class="sc">%&gt;%</span> </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Mean=</span><span class="fu">mean</span>(Mean)) <span class="sc">%&gt;%</span> </span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Taxon) <span class="sc">%&gt;%</span> </span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Mean=</span><span class="fu">mean</span>(Mean)) <span class="sc">%&gt;%</span> </span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(Mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  Taxon  Mean
  &lt;int&gt; &lt;dbl&gt;
1     1  56.6
2     0  55.8</code></pre>
</div>
</div>
<p>We declare the input data set in the first line, and then pipe to <code>group_by()</code> in line 2. The output of <code>group_by()</code> is then *piped to <code>summarise()</code>, and so on. There are two important things to note here:</p>
<ol type="1">
<li>We do not declare the data object inside of each function. The pipe command does this for us.</li>
<li>We still have to assign the output to an object if we want to retain it. For example, we might change line 1 to <code>wtMean&lt;-RFallo %&gt;%</code>. We didn’t redirect the command to an object, so it simply output to your R Console.</li>
</ol>
<p>Compared to the earlier example, it is clear that the pipe can make for cleaner, more concise code with fewer objects added to the environment. also avoids potential for bugs in our program. Imagine if we mis-spelled ‘Taxon’ in our second line by accidentally pressing ‘s’ along with ‘x’. Compare the output:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">group_by</span>(Fallo,Taxon,Scenario,Nutrients)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">group_by</span>(X,Tasxon,Scenario)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `group_by()`:
! Must group by variables found in `.data`.
✖ Column `Tasxon` is not found.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">group_by</span>(X,Taxon)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">summarise</span>(X, <span class="at">Mean=</span><span class="fu">mean</span>(Total), <span class="at">SD=</span><span class="fu">sd</span>(Total))</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">arrange</span>(X,<span class="fu">desc</span>(Mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 3
  Taxon  Mean    SD
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
1 japon  56.4 10.4 
2 bohem  56.3  9.54</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>Fallo <span class="sc">%&gt;%</span> </span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Taxon,Scenario,Nutrients) <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Tasxon,Scenario) <span class="sc">%&gt;%</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Taxon) <span class="sc">%&gt;%</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Mean=</span><span class="fu">mean</span>(Total), <span class="at">SD=</span><span class="fu">sd</span>(Total)) <span class="sc">%&gt;%</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(Mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in `group_by()`:
! Must group by variables found in `.data`.
✖ Column `Tasxon` is not found.</code></pre>
</div>
</div>
<p>In both cases we get an error, but in one case we still calculate the means and sd of the two species.</p>
<blockquote class="blockquote">
<p>A bug that produces no output is much less dangerous than an error that gives an output. Why?</p>
</blockquote>
</section>
</section>
<section id="join-datasets" class="level2">
<h2 class="anchored" data-anchor-id="join-datasets">Join Datasets</h2>
<p>The <code>dplyr</code> package has some handy <code>joing_</code> tools to combine data sets. There are four main ways to join data sets. I find it helpful to think of Venn diagrams when I’m trying to remember what all of these functions do. You can get more information on these with the help command <code>?join</code> after loading the dplyr library, but here is a quick overview. For each of these, imagine a Venn diagram with two datasets: <span class="math inline">\(X\)</span> as a circle on the left side and <span class="math inline">\(Y\)</span> as a circle on the right side. The rows we choose to combine from the two datasets depend on one or more identifying columns that we can define (e.g.&nbsp;sample ID, date, time).</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/joinVenn.png" class="img-fluid figure-img" alt="Venn diagram of datasets"></p>
<figcaption>Venn diagram of datasets</figcaption>
</figure>
</div>
<ol type="1">
<li><p><code>left_join()</code> - Keep all rows of <code>X</code> and add matching rows from <code>Y</code>. Any rows in <code>Y</code> that don’t match <code>X</code> are excluded.</p></li>
<li><p><code>right_join()</code> - The reverse of <code>left_join()</code></p></li>
<li><p><code>inner_join()</code> - Only keep rows that are common to <strong>both</strong> <code>X</code> <strong>AND</strong> <code>Y</code>. Remove any rows that don’t match.</p></li>
<li><p><code>full_join()</code> - Keep <strong>any</strong> columns that are in <strong>either</strong> <code>X</code> <strong>OR</strong> <code>Y</code>. Add <code>NA</code> for missing data in either of the columns for rows that don’t match.</p></li>
</ol>
<p>To try these out, we can create a couple of quick data sets and compare the output. For each case, note the addition of NA for missing data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">data.frame</span>(<span class="at">ID=</span><span class="fu">c</span>(<span class="st">"A"</span>,<span class="st">"C"</span>),<span class="at">Xdat=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>Y<span class="ot">&lt;-</span><span class="fu">data.frame</span>(<span class="at">ID=</span><span class="fu">c</span>(<span class="st">"B"</span>,<span class="st">"C"</span>),<span class="at">Ydat=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">3</span>))</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat
1  A    1
2  C    3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>Y</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Ydat
1  B    2
2  C    3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">left_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat Ydat
1  A    1   NA
2  C    3    3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">right_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat Ydat
1  C    3    3
2  B   NA    2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">inner_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat Ydat
1  C    3    3</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">full_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat Ydat
1  A    1   NA
2  C    3    3
3  B   NA    2</code></pre>
</div>
</div>
</section>
<section id="wide-vs-long-data" class="level2">
<h2 class="anchored" data-anchor-id="wide-vs-long-data"><em>Wide</em> vs <em>Long</em> Data</h2>
<p>Most of the data examples we’ve looked at are in the ‘wide’ format, where we have a single individual as a row, and multiple measurements as separate columns.</p>
<p>However, there are many cases where we may want to reorganize our data into the ‘long’ format, where each row is an individual observation. Many statistical models use this format, and it’s also useful for visualizations.</p>
<p>Here is one very handy example that I use all the time. If we have a bunch of columns of observations and we want to generate plots quickly (e.g.&nbsp;frequency distributions), we can convert the data to the <em>long</em> format and then use the <code>facet_</code> functions from <code>ggplot2</code> to create separate plots for each of the original columns containing the observations.</p>
<section id="pivot_longer" class="level3">
<h3 class="anchored" data-anchor-id="pivot_longer"><code>pivot_longer</code></h3>
<p>The <code>pivot_longer()</code> function in the <code>tidyr</code> library converts data from <em>wide to long</em>. We use the <code>cols=</code> parameter to specify the data columns, then <code>names_to=</code> specifies the column name containing the parameter, and the <code>cols_to=</code> specifies the column name of the values.</p>
<p>This is a bit confusing to read, but it’s easier to understand if you compare the output with the <code>full_join</code> function in the previous section.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>LongData<span class="ot">&lt;-</span><span class="fu">full_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols=</span><span class="fu">c</span>(<span class="st">"Xdat"</span>,<span class="st">"Ydat"</span>),</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to=</span><span class="st">"Measurement"</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to=</span><span class="st">"Value"</span>)</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>LongData</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  ID    Measurement Value
  &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt;
1 A     Xdat            1
2 A     Ydat           NA
3 C     Xdat            3
4 C     Ydat            3
5 B     Xdat           NA
6 B     Ydat            2</code></pre>
</div>
</div>
<p>This is the <em>long</em> format. Compare to the <em>wide</em> format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">full_join</span>(X,Y,<span class="at">by=</span><span class="st">"ID"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  ID Xdat Ydat
1  A    1   NA
2  C    3    3
3  B   NA    2</code></pre>
</div>
</div>
<p>Note how there is only one column of data values, with the <code>Measurement</code> column indicating which measurement the value belongs to, and the <code>ID</code> column is repeated for each measurement.</p>
<p>This is why the <em>long</em> data format is sometimes called the <em>repeated measures</em> data format.</p>
</section>
<section id="pivot_wider" class="level3">
<h3 class="anchored" data-anchor-id="pivot_wider"><code>pivot_wider</code></h3>
<p>The <code>pivot_wider()</code> function does the reverse. This time, we specify the column that contains the values with <code>values_from=</code> and the corresponding column names with <code>names_from=</code>. This should recover the original data set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>WideData<span class="ot">&lt;-</span>LongData <span class="sc">%&gt;%</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">values_from=</span>Value,</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">names_from=</span>Measurement)</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>WideData</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 3
  ID     Xdat  Ydat
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt;
1 A         1    NA
2 C         3     3
3 B        NA     2</code></pre>
</div>
</div>
<p>Note the slight difference in output.</p>
<blockquote class="blockquote">
<p><strong>Question</strong>: Can you explain why this output is different than the original?</p>
</blockquote>
<p><strong>Answer</strong>: The original was a <code>data.frame</code> object, but the output of <code>pivot_longer()</code> and <code>pivot_wider()</code></p>
</section>
</section>
<section id="missing-data" class="level2">
<h2 class="anchored" data-anchor-id="missing-data">Missing Data</h2>
<p>So far we have worked on a pristine data set that has already been edited for errors. More often datasets will contain missing values.</p>
<section id="na-and-na.rm" class="level3">
<h3 class="anchored" data-anchor-id="na-and-na.rm"><code>NA</code> and <code>na.rm()</code></h3>
<p>As we have already seen, the R language uses a special object <code>NA</code> to denote missing data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>Vec<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="cn">NA</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>Vec</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  1  2  3 NA  5  6  7</code></pre>
</div>
</div>
<p>When a function is run on a vector or other object containing NA, the function will often return NA or give an error message:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(Vec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p>This is by design, because it is not always clear what <code>NA</code> means. Should these data be removed from consideration, or is it a zero to be included in calculations and statistical models? Many functions in R include an na.rm parameter that is set to <code>FALSE</code> by default. Setting it to true tells the function to ignore the <code>NA</code> (i.e., remove it from the calculation).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(Vec, <span class="at">na.rm=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
</section>
<section id="na-vs-0" class="level3">
<h3 class="anchored" data-anchor-id="na-vs-0"><code>NA</code> vs <code>0</code></h3>
<p>A common mistake students make is to put 0 for missing data. This can be a big problem when analyzing the data since the calculations are very different.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>Vec1<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="cn">NA</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(Vec1, <span class="at">na.rm=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>Vec2<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">0</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(Vec2, <span class="at">na.rm=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3.428571</code></pre>
</div>
</div>
<p>However, there are many cases where <code>NA</code> <em>does</em> represent a zero for purposes of calculation or statistical analysis. There is no simple rule to follow here. The best decision will depend on the specific details of the biological data and the assumptions of the particular calculation or statistic.</p>
</section>
<section id="is.na" class="level3">
<h3 class="anchored" data-anchor-id="is.na"><code>is.na()</code></h3>
<p>In large datasets you might want to check for missing values. Let’s simulate this in our <em>FallopiaData.csv</em> dataset.</p>
<p>To set up a test data frame, randomly select 10 rows and replace the value for ‘Total’ with NA.</p>
<blockquote class="blockquote">
<p><strong>Question</strong>: Can you remember how to do this, from the <em>R Fundamentals</em> Chapter?</p>
</blockquote>
<p><strong>Answer</strong>: There are many ways that you could approach this. One way is to first randomly select from a vector representing the row number in our data frame. Then, replace the values of these particular observations with <code>NA</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>X<span class="ot">&lt;-</span><span class="fu">sample</span>(<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Fallo)),<span class="dv">10</span>,<span class="at">replace=</span>F)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>Fallo<span class="sc">$</span>Total[X]<span class="ot">&lt;-</span><span class="cn">NA</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>Fallo<span class="sc">$</span>Total</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 67.06 50.22 61.08 41.71 41.81 48.27 55.42    NA 53.53 45.89 59.02 57.66
 [13] 48.98 35.97 43.28 52.27    NA 44.61 59.13 58.97 55.36 31.46 43.46 44.65
 [25] 59.69 60.82 57.21 34.09 58.57    NA 63.18 59.88 54.09 55.27 61.31 53.56
 [37] 52.66 64.71 61.06 45.34 64.20 57.50 68.55 49.55    NA 54.06 66.60 74.82
 [49] 53.71 49.75 58.45 66.06 67.01 70.41    NA 63.43    NA 47.50 61.79 54.96
 [61] 48.99 52.01    NA    NA 42.47 46.18 62.56 54.36 69.54 75.91 56.34 64.97
 [73] 60.71 57.80 41.72 67.44 58.78 77.74 65.68 58.42 55.35 50.28 55.04 39.56
 [85] 71.07 45.23 57.20 67.70 52.46 60.86 62.19 65.53 48.19 60.89 48.13 60.37
 [97] 67.86 56.40 49.13 56.11 49.78 69.00 65.40    NA 63.08 60.93 29.54 49.12
[109] 68.73 31.90 69.88 69.48 47.88    NA 58.13 50.51 54.83 66.80 50.31 56.12
[121] 62.96 78.80 64.25</code></pre>
</div>
</div>
<p>Use <code>is.na()</code> to check for missing values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is.na</span>(Fallo<span class="sc">$</span>Total)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
 [13] FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [25] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
 [37] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE
 [49] FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE
 [61] FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [73] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [85] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
 [97] FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE
[109] FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
[121] FALSE FALSE FALSE</code></pre>
</div>
</div>
<p>Note that the output is a vector of True/False. Each cell corresponds to a value of ‘Total’ with TRUE indicating missing values. This is an example of a boolean variable, which has some handy properties in R.</p>
<p>First, we can use it as an index. For example, let’s see which pots have missing ‘Total’ values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>Missing<span class="ot">&lt;-</span><span class="fu">is.na</span>(Fallo<span class="sc">$</span>Total)</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>Fallo<span class="sc">$</span>PotNum[Missing]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]   9  20  36  55  68  70  78  79 126 138</code></pre>
</div>
</div>
<p>Another handy trick to check for missing values is to sum the vector:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">is.na</span>(Fallo<span class="sc">$</span>Total))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10</code></pre>
</div>
</div>
<p>This takes advantage of the fact that the Boolean TRUE/FALSE variable is equivalent to the binary 1/0 values. By default, R treats <code>TRUE</code> as 1 and <code>FALSE</code> as 0 if used in any mathematical operation.</p>
</section>
</section>
<section id="naughty-data" class="level2">
<h2 class="anchored" data-anchor-id="naughty-data">Naughty Data</h2>
<p>Naughty data contain the same information as a standard row x column (i.e.&nbsp;2-dimensional) data frame but break tone or more of the following best practices.</p>
<ol type="1">
<li>Each cell contains a single value</li>
<li>Each variable must have its own column</li>
<li>Each observation must have its own row</li>
</ol>
<p>Naughty data are very common in biology, this often occurs when either the collector(s) were not familiar with these best practices, or decisions are made to favour human readibility and usability rather than computer interpretability.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./images/Wu_Fig2.jpg" class="img-fluid figure-img" alt="Examples of Naughty Data from Wu et al. (2022), shown on the left side. The right side shown well-behaved data, with one observation per row, and column names as the first row."></p>
<figcaption>Examples of Naughty Data from Wu et al.&nbsp;(2022), shown on the left side. The right side shown well-behaved data, with one observation per row, and column names as the first row.</figcaption>
</figure>
</div>
<p>Naughty data can be very time consuming to fix, but regular expressions can make this a bit easier, as discussed in the <em>Regular Expressions</em> Chapter.</p>
</section>
<section id="dates" class="level2">
<h2 class="anchored" data-anchor-id="dates">Dates</h2>
<p>As biologists, we often work with dates or time. We may want to analyze the date a sample was collected, or a measurement was taken. But dates are often encoded in formats that don’t fit neatly into the usual data types that we’ve worked with so far. The <code>lubridate</code> package provides a convenient framework for switching between human-readable dates and mathematical relationships among them – for example, the number of days, minutes, or seconds between two time points.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It can be convenient to automatically include the current date or time, especially when you are producing reports in R.</p>
<p>We can get the date as a <code>date</code> object in the <em>year-month-day</em> format,</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">today</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2024-01-29"</code></pre>
</div>
</div>
<p>And we can get the date along with the time as a <code>datetime</code> object, in the <em>year-month-day hour:minute:second timezone</em> format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">now</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2024-01-29 20:45:14 EST"</code></pre>
</div>
</div>
<p>This is an important distinction, because the <code>datetime</code> object extends on the <code>date</code> object to include hours, minutes, seconds, and time zone.</p>
<section id="run-time" class="level3">
<h3 class="anchored" data-anchor-id="run-time">Run Time</h3>
<p>We can use a <code>datetime</code> object to track the <em>run time</em> of our code. The <strong>run time</strong> is how long it takes to run a particular program. We first create an object to store the computer clock before the program runs, then we subtract that object from the computer clock after the program finishes:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>Before<span class="ot">&lt;-</span><span class="fu">now</span>()</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">1000</span>){</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rpois</span>(<span class="dv">10</span>,<span class="at">lambda=</span><span class="dv">10</span>)<span class="sc">^</span><span class="fu">rnorm</span>(<span class="dv">10</span>)</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a><span class="fu">now</span>()<span class="sc">-</span>Before</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 0.0129571 secs</code></pre>
</div>
</div>
<p>Note that your run time may be different, depending on the specifications of your computer and other processes that are running on it.</p>
<p>This technique can be useful for estimating the run time of a large program. Specifically, we run a fraction of the data and/or loop iterations and multiply to get an estimate the run time for the full data set.</p>
<blockquote class="blockquote">
<p><strong>Question</strong>: If you add a <code>print()</code> function to print out the result in each iteration of the for loop, how much does this slow down the run time?</p>
</blockquote>
<p><strong>Answer</strong>: Try it to find out!</p>
<p>For a small program like this, it’s not a problem to ad a few fractions of a second. But this can scale up exponentially in ‘big data’ applications. For example, imagine you are sequencing a human genome at 100× coverage, which might typically involve one a billion sequences, each 300 bases long. You want to write a program to assemble all of the sequences into a full genome. How much time would it add to your program run time if you add a 1 millisecond step for each sequence, or each base pair in the experiment?</p>
<p>For each sequence, it would be <span class="math inline">\(10^9sequences \times 0.001 s = 10^6 s\)</span>, or about 11.5 days! For each base pair, multiply by 300 – it would take more than 10 years!</p>
<p>These are problems of scale that don’t matter so much when you are starting out, but developing an efficiency mindset early, it can pay off when we move on to bigger data projects.</p>
<p>Next, we’ll explore the <code>date</code> and <code>datetime</code> objects in more detail.</p>
</section>
<section id="date-objects" class="level3">
<h3 class="anchored" data-anchor-id="date-objects"><code>date</code> Objects</h3>
<p>Human-readable dates come in many different forms, which we can encode as strings. Here are some examples that we might see for encoding the end-date of the most recent 5,126-year-long Mesoamerican calendar used by the ancient Maya civilization:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>Date2<span class="ot">&lt;-</span><span class="st">"2012-12-21"</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>Date3<span class="ot">&lt;-</span><span class="st">"21.12.2012"</span></span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>Date4<span class="ot">&lt;-</span><span class="st">"Dec 21, 2012"</span></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>Date5<span class="ot">&lt;-</span><span class="st">"21 December, 2012"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>lubridate</code> package has a number of related functions that correspond to the order of entry – <code>d</code> for day, <code>m</code> for month, and <code>y</code> for year:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ymd</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dmy</span>(Date3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mdy</span>(Date4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dmy</span>(Date5)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21"</code></pre>
</div>
</div>
<p>Notice the flexibility here! Some dates have dashes, dots, spaces, and commas! Yet, all are easily converted to a common object type. On the surface, these objects look like simple strings, but compare the structure of the <code>date</code> object with its original input string:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> chr "2012-12-21"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(<span class="fu">ymd</span>(Date2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> Date[1:1], format: "2012-12-21"</code></pre>
</div>
</div>
<p>Notice how the first is a simple <code>chr</code> character object, whereas the second is a <code>Date</code> object. The date object can be treated as a numeric variable, that outputs as a readable date. For example, what if we want to know what the date 90 days before or after?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">ymd</span>(Date2)<span class="sc">-</span><span class="dv">90</span>,<span class="fu">ymd</span>(Date2)<span class="sc">+</span><span class="dv">90</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-09-22" "2013-03-21"</code></pre>
</div>
</div>
<p>When using <code>date</code> objects, R will even account for the different number of days in each month, and adjust for leap years!</p>
<p>As with any function in R, we can apply <code>ymd()</code> to a column in a data frame or any other vector of strings:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>DateVec<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"2012-12-1"</span>,<span class="st">"2012-12-20"</span>,</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">"2012-12-23"</span>, <span class="st">"2012-11-05"</span>)</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ymd</span>(DateVec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-01" "2012-12-20" "2012-12-23" "2012-11-05"</code></pre>
</div>
</div>
<p>The elements must have the have the same order as the function, but surprisingly they don’t have to have the same format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>MixVec<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">"2012-12-11"</span>,<span class="st">"12-20-2012"</span>,</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>          <span class="st">"2012, December 21"</span>, <span class="st">"2013, Nov-11"</span>)</span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ymd</span>(MixVec)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: 1 failed to parse.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-11" NA           "2012-12-21" "2013-11-11"</code></pre>
</div>
</div>
<p>Note the warning and the resulting output. The first, third, and fourth elements are converted, even though they have different formats. The second element is replaced with <code>NA</code> because the order is not year, month, day, as required by <code>ymd()</code>.</p>
<p>Because these objects are numeric, we can also use them for plotting:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="fu">ymd</span>(DateVec)),</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">binwidth=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="datascience_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Question</strong>: What do you notice about the x-axis? How would it differ if we used strings instead?</p>
</blockquote>
<p><strong>Answer</strong>: Try to run the above with DateVec as a string instead of <code>date</code>.</p>
<p>You should get an error, because you can’t bin values of a character variable to generate a frequency histogram. This shows us that R is able to treat the <code>date</code> object as a continuous variable, spacing the bins based on the number of days separating them.</p>
</section>
<section id="datetime" class="level3">
<h3 class="anchored" data-anchor-id="datetime"><code>datetime</code></h3>
<p>The <code>datetime</code> object adds a time element to the date. Just as there are different functions to translate different date formats, there are different <code>datetime</code> functions. Each <code>datetime</code> function starts with one of the three <code>date</code> functions we used earlier, but then adds time elements after an underscore <code>_</code> to define nine different functions. Here are a few examples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mdy_h</span>(<span class="st">"Dec 21, 2012 -- 10am"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21 10:00:00 UTC"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ymd_hm</span>(<span class="st">"2012-12-21, 08:30"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21 08:30:00 UTC"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dmy_hms</span>(<span class="st">"21 December, 2012; 11:59:59"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2012-12-21 11:59:59 UTC"</code></pre>
</div>
</div>
</section>
<section id="extracting-components" class="level3">
<h3 class="anchored" data-anchor-id="extracting-components">Extracting Components</h3>
<p>We can extract components from <code>date</code> and <code>datetime</code> objects. For example, we can extract only the year:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">year</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2012</code></pre>
</div>
</div>
<p>or the month:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">month</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
</div>
<p>We have several options for days.</p>
<p>First, we can get the day of the year, also known as the Julian day:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">yday</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 356</code></pre>
</div>
</div>
<p>Or the day of the month</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mday</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 21</code></pre>
</div>
</div>
<p>or the day of the week (rather than month) using <code>wday()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wday</span>(Date2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>We can use the <code>label=T</code> parameter rather than number, to get the name of the specific month or day, rather than the number:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a><span class="fu">month</span>(Date2, <span class="at">label=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Dec
12 Levels: Jan &lt; Feb &lt; Mar &lt; Apr &lt; May &lt; Jun &lt; Jul &lt; Aug &lt; Sep &lt; ... &lt; Dec</code></pre>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wday</span>(Date2, <span class="at">label=</span>T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Fri
Levels: Sun &lt; Mon &lt; Tue &lt; Wed &lt; Thu &lt; Fri &lt; Sat</code></pre>
</div>
</div>
</section>
<section id="categorical-dates" class="level3">
<h3 class="anchored" data-anchor-id="categorical-dates">Categorical Dates</h3>
<p>We’ve seen above how dates are more like <code>numeric</code> objects rather than <code>strings</code>, but we can also treat dates as categorical data.</p>
<p>One example that is becoming more and more common in biology is the analysis of data from data loggers, which automatically save observations over time. Think of climate stations that measure temperature and precipitation as a common example. Another example might be location information of a study organism using image analysis or PIT tags (i.e.&nbsp;Passive Integrated Transponders).</p>
<p>In many cases, the timescale of collection is much shorter than what we need for our analysis. We end up with too much data! Luckily, <code>dplyr</code> with <code>lubridate</code> offer a great combination for summarizing these data.</p>
<section id="lubridate-dplyr" class="level4">
<h4 class="anchored" data-anchor-id="lubridate-dplyr"><code>lubridate</code> + <code>dplyr</code></h4>
<p>Here’s an example that takes advantage that takes advantage of our ability to treat components of <code>datetime</code> objects as categorical variables as well as continuous variables.</p>
<p>Imagine we have observations taken every minute, and we just want to calculate the average for each hour . To see how to do this, we will generate a toy data set in R using the tibble function, and then assigning column names: <code>DayTime</code> for the day-time object and <code>Obs</code> for the observation (i.e.&nbsp;measurement):</p>
<p>First, we create the imaginary dataset, using the replicate function and a random number generator:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a>TestData<span class="ot">&lt;-</span><span class="fu">tibble</span>(</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">DayTime=</span><span class="fu">now</span>()<span class="sc">+</span><span class="fu">minutes</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">359</span>),<span class="dv">100</span>)),</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Obs=</span><span class="fu">rnorm</span>(<span class="dv">36000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can calculate the hourly average by piping our <code>TestData</code> tibble (above) through a group_by and then a <code>summarise()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>TestData <span class="sc">%&gt;%</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(<span class="fu">yday</span>(DayTime),<span class="fu">hour</span>(DayTime)) <span class="sc">%&gt;%</span></span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Mean=</span><span class="fu">mean</span>(Obs))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'yday(DayTime)'. You can override using the
`.groups` argument.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 7 × 3
# Groups:   yday(DayTime) [2]
  `yday(DayTime)` `hour(DayTime)`      Mean
            &lt;dbl&gt;           &lt;int&gt;     &lt;dbl&gt;
1              29              20  0.0161  
2              29              21 -0.0170  
3              29              22  0.0111  
4              29              23 -0.0108  
5              30               0 -0.0159  
6              30               1 -0.0283  
7              30               2 -0.000328</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><strong>Question</strong> Why do we include <code>yday()</code> in the group_by function?</p>
</blockquote>
<p><strong>Answer</strong>: Remove <code>yday(DayTime)</code> and compare the output.</p>


</section>
</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>R Crash Course for Biologists</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p><a href="https://EcoEvoGeno.org/">Robert I Colautti</a></p>
</div>
  </div>
</footer>




</body></html>